### PR TITLE
feat: Server side swift

### PR DESCRIPTION
**Pull request by eriadam**
As described in the [feature request](https://github.com/algolia/algoliasearch-client-swift/issues/660), the aim of this PR is to add support for server side swift.

**This is work in progress.**

TODOs:

- [x] CommonCrypto vs swift-crypto
- [x] Updating/runnig tests for Linux target
- [x] Examples
- [x] Linux CI tests
---

### FoundationNetworking

Network related functionality was moved out of Foundation to the new FoundationNetworking module. So this needs a conditional import.

```swift
#if os(Linux)
import FoundationNetworking
#endif
```

### CommonCrypto

This is not part of the open source distribution. Instead, the recommendation is to use [swift-crypto](https://github.com/apple/swift-crypto), which _"is an open-source implementation of a substantial portion of the API of Apple [CryptoKit](https://developer.apple.com/documentation/cryptokit)"._

The only part I found using this is the generation of secured API key using HMAC. I managed to solve this bit by adding swift-crypto as a dependency and rewriting the generation using the other API.

### OperationLauncher

It seems that on Linux the `OperationQueue` used in the `OperationLauncher` does not get an underlying queue, or gets something wrong. This results in operations not being started after adding them and sync operations not being finished. 

```swift
class OperationLauncher {

  let queue: OperationQueue

  init(queue: OperationQueue) {
    self.queue = queue
#if os(Linux)
    self.queue.underlyingQueue = DispatchQueue
      .global(qos: DispatchQoS.QoSClass.background)
#endif
  }
```

### CODE REVIEW

**Comment by eriadam** [comment]
Hey @VladislavFitz, how would I have access to the test results?

**Comment by VladislavFitz** [comment]
Hi @eriadam, 

Sorry for late response.
Unfortunately, we still didn't finish the flow allowing the contributors to pass the tests, but we are working on it. 
The CI failed because there is a merge conflict between your fork and the master branch.
I tested it locally and all tests passed ✅ 
So, the next steps: 
1) I'll make a PR to you fork with the necessary changes
2) You merge the PR to your `server-side-swift` branch
3) The PR becomes ready for review and must succeed the CI check
4) We pass the last check and merge your PR

Thanks again for your contribution! 🙌

**Comment by VladislavFitz** [comment]
Hi @eriadam,
I made the necessary changes, but I cannot merge them to your `server-side-swift` branch. 
Could you allow me to make changes as described in [this guide](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)?

**Comment by VladislavFitz** [comment]
Finally, all checks passed including our CI ✅ 
@eriadam if it is ok for you, we can include this PR in the next release 🚀

**Comment by eriadam** [comment]
> Finally, all checks passed including our CI ✅
> @eriadam if it is ok for you, we can include this PR in the next release 🚀

Awesome! 
I would add a few examples, especially with [Vapor](https://vapor.codes) threading is interesting. I will do that tonight.

**Comment by eriadam** [comment]
I added some examples for using this with Vapor and NIO: https://github.com/algolia/api-clients-playground/pull/32

File: Sources/AlgoliaSearchClient/Helpers/Crypto/HMAC.swift [hunk 1]
```diff
@@ -6,6 +6,8 @@
 //
 
 import Foundation
+#if os(Linux)
+#else
```
Comment (spinach) [inline Sources/AlgoliaSearchClient/Helpers/Crypto/HMAC.swift:10]: Is this still needed?

File: Sources/AlgoliaSearchClient/Helpers/Crypto/String+HMAC.swift [hunk 2]
```diff
@@ -6,13 +6,24 @@
...
 import CommonCrypto
+#endif
 
 extension String {
-
-  func hmac(algorithm: HmacAlgorithm, key: String) -> String {
+  
+  func hmac256(withKey key: String) -> String {
+    #if os(Linux)
+    let key = SymmetricKey(data: key.data(using: .utf8)!)
```
Comment (spinach) [inline Sources/AlgoliaSearchClient/Helpers/Crypto/String+HMAC.swift:20]: Any way we can avoid for unwrapping in this function? I see 2

**Review by spinach — APPROVED** [review]

File: Sources/AlgoliaSearchClient/Helpers/Crypto/String+HMAC.swift (hunk 2 above)
Comment (eriadam) [reply Sources/AlgoliaSearchClient/Helpers/Crypto/String+HMAC.swift:20]: It is safe to force-unwrap _unicode_ string-to-data transformation results.

But this should work as well.

```
    let key = SymmetricKey(data: Data(key.utf8))
    let digest = Array(HMAC<SHA256>.authenticationCode(for: Data(key.utf8), using: key))
```

Comment (VladislavFitz) [reply Sources/AlgoliaSearchClient/Helpers/Crypto/String+HMAC.swift]: Fixed!

**Review by VladislavFitz — APPROVED** [review]

**Pull request merged**